File: backend/src/main/java/com/imedba/modules/book/entity/Book.java
package com.imedba.modules.book.entity;

import com.imedba.common.entity.BaseEntity;
import com.imedba.modules.course.entity.BusinessUnit;
import jakarta.persistence.Column;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Entity;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Builder.Default;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.SQLRestriction;

/**
 * Libro del catálogo editorial. Mantiene stock plano (stock_quantity + branch)
 * por decisión del ERD simplificado. Si en el futuro se necesita stock por sede,
 * se migra a {@code book_stocks} y se deja {@code stock_quantity} como total
 * calculado.
 */
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "books")
@SQLDelete(sql = "UPDATE books SET deleted_at = NOW() WHERE id = ?")
@SQLRestriction("deleted_at IS NULL")
public class Book extends BaseEntity {

    @Column(name = "name", nullable = false, length = 200)
    private String name;

    /**
     * Unidad en la que se ofrece el libro (V036). Filtra el selector de libros de la
     * inscripción por la unidad del curso: un alumno de Residencias no debería poder
     * sumar el libro de PREMA a su matrícula. NULL = disponible en todas.
     */
    @Enumerated(EnumType.STRING)
    @Column(name = "business_unit", length = 30)
    private BusinessUnit businessUnit;

    @Column(name = "code", length = 20)
    private String code;

    @Column(name = "specialty", length = 100)
    private String specialty;

    @Column(name = "format", length = 50)
    private String format;

    @Column(name = "edition", length = 50)
    private String edition;

    @Column(name = "pages")
    private Integer pages;

    @Column(name = "sale_price", nullable = false, precision = 12, scale = 2)
    private BigDecimal salePrice;

    @Default
    @Column(name = "student_discount_pct", nullable = false, precision = 5, scale = 2)
    private BigDecimal studentDiscountPct = new BigDecimal("30.00");

    /**
     * % del precio de venta que va a la torta de autorías; los
     * royalty_percentage de book_authors reparten este pool (doc 09 §3.4).
     */
    @Default
    @Column(name = "royalty_pool_pct", nullable = false, precision = 5, scale = 2)
    private BigDecimal royaltyPoolPct = new BigDecimal("10.00");

    @Column(name = "cost_per_unit", precision = 12, scale = 2)
    private BigDecimal costPerUnit;

    @Default
    @Column(name = "stock_quantity", nullable = false)
    private Integer stockQuantity = 0;

    @Column(name = "branch", length = 100)
    private String branch;

    @Default
    @Column(name = "is_active", nullable = false)
    private Boolean active = Boolean.TRUE;
}

File: frontend/src/types/book.ts
import type { Instant, UUID } from './common'
import type { BusinessUnit } from './course'

// Refleja com.imedba.modules.book.dto.BookAuthorResponse
export interface BookAuthor {
  authorId:          UUID
  firstName:         string
  lastName:          string
  royaltyPercentage: number   // 0–100
}

// Refleja com.imedba.modules.book.dto.BookResponse
export interface Book {
  id:                 UUID
  name:               string
  /** Unidad en la que se ofrece (V036). null = todas. */
  businessUnit:       BusinessUnit | null
  code:               string | null
  specialty:          string | null
  format:             string | null
  edition:            string | null
  pages:              number | null
  salePrice:          number          // requerido, ≥ 0
  studentDiscountPct: number | null   // % off alumnos
  royaltyPoolPct:     number          // % de la venta destinado al pool de autorías (default 10)
  costPerUnit:        number | null
  stockQuantity:      number | null
  branch:             string | null
  active:             boolean
  authors:            BookAuthor[]
  createdAt:          Instant
  updatedAt:          Instant
}

// Refleja BookAuthorRequest
export interface BookAuthorRequest {
  authorId:          UUID
  royaltyPercentage: number   // 0–100
}

// Refleja BookCreateRequest
export interface BookCreateRequest {
  name:                string                 // required, max 255
  code?:               string | null          // max 50
  specialty?:          string | null          // max 100
  format?:             string | null          // max 50
  edition?:            string | null          // max 50
  pages?:              number | null          // ≥ 0
  salePrice:           number                 // required, ≥ 0
  studentDiscountPct?: number | null          // ≥ 0
  royaltyPoolPct?:     number | null          // 0–100, default 10
  costPerUnit?:        number | null          // ≥ 0
  stockQuantity?:      number | null          // ≥ 0
  branch?:             string | null          // max 50
  authors?:            BookAuthorRequest[]
}

// Refleja BookUpdateRequest (sin authors — se gestionan por endpoints aparte;
// sin stockQuantity — se ajusta por PUT /books/{id}/stock para no pisar las ventas).
export interface BookUpdateRequest {
  name?:               string
  code?:               string | null
  specialty?:          string | null
  format?:             string | null
  edition?:            string | null
  pages?:              number | null
  salePrice?:          number
  studentDiscountPct?: number | null
  royaltyPoolPct?:     number | null   // 0–100
  costPerUnit?:        number | null
  branch?:             string | null
  active?:             boolean
}

// Refleja BookStockUpdateRequest — ajuste manual del stock con motivo opcional.
export interface BookStockUpdateRequest {
  stockQuantity: number
  reason?:       string | null
}
